twill-yarn/src/main/java/org/apache/twill/filesystem/FileContextLocationUtil.java

```diff
@@ -35,13 +34,26 @@
   // To check whether logical URI is needed, hdfs.HAUtil class is used. But the class is meant for internal purposes,
   // and in Hadoop 2.8, the method was renamed from "isLogicalUri" to "useLogicalUri". So resolve to the
   // correct method.
+  // In Hadoop 3.0 and above, HAUtil is not available, instead HAUtilClient should be used
   private static final MethodHandle HA_UTIL_USE_LOGICAL_URI_HANDLE;
 
-  private static MethodHandle lookupInHAUtil(final String methodName)
-      throws NoSuchMethodException, IllegalAccessException {
-    return MethodHandles.publicLookup()
-        .findStatic(HAUtil.class, methodName,
-            MethodType.methodType(boolean.class, new Class[]{Configuration.class, URI.class}));
+  private static MethodHandle lookupInHAUtil(String methodName)
+    throws NoSuchMethodException, IllegalAccessException {
+    return MethodHandles.publicLookup().findStatic(
+      HAUtil.class, methodName,
```

Comment:
Use the string class name and use `loadClass` such that it gives ClassNotFoundException instead of NoClassDefError. Also, this and the next new method can be refactored to be one.